File: src/main/java/com/assetmanagement/maintenance/service/MaintenanceService.java
package com.assetmanagement.maintenance.service;

import com.assetmanagement.maintenance.dto.request.*;
import com.assetmanagement.maintenance.dto.response.MaintenanceHistoryResponse;
import com.assetmanagement.maintenance.dto.response.MaintenanceIssueResponse;
import com.assetmanagement.maintenance.dto.response.MaintenanceWorkOrderResponse;

import java.util.List;

public interface MaintenanceService {

    /** Employee reports a maintenance issue for an asset. */
    MaintenanceIssueResponse reportIssue(ReportIssueRequest request);

    /** Get a single issue with all its work order history. */
    MaintenanceIssueResponse getIssueById(Long issueId);

    /** Get all maintenance issues in the system. */
    List<MaintenanceIssueResponse> getAllIssues();

    /**
     * Manager assigns a technician to an issue.
     * Creates a new WorkOrder attempt.
     */
    MaintenanceWorkOrderResponse createWorkOrder(Long issueId, AssignTechnicianRequest request);

    /** Legacy alias for assignTechnician. */
    MaintenanceIssueResponse assignTechnician(Long issueId, AssignTechnicianRequest request);

    /** Technician starts the work order. Sets status = IN_PROGRESS and Asset status = UNDER_MAINTENANCE. */
    MaintenanceWorkOrderResponse startWorkOrder(Long workOrderId);

    /**
     * Technician accepts or rejects the assigned work order (full object response).
     */
    MaintenanceWorkOrderResponse respondToWorkOrder(Long workOrderId, RespondWorkOrderRequest request);

    /** Technician rejects the work order with a reason string. */
    MaintenanceWorkOrderResponse rejectWorkOrder(Long workOrderId, RejectWorkOrderRequest request);

    /**
     * Technician submits repair completion.
     */
    MaintenanceWorkOrderResponse completeRepair(Long workOrderId, CompleteRepairRequest request);

    /** Technician marks work order as NOT_REPAIRABLE with a reason. */
    MaintenanceWorkOrderResponse markNotRepairable(Long workOrderId, NotRepairableRequest request);

    /**
     * Manager decides on a NOT_REPAIRABLE work order (RETIRE or REPLACE).
     */
    MaintenanceWorkOrderResponse applyWorkOrderDecision(Long workOrderId, ManagerDecisionRequest request);

    /**
     * Manager decides on a NOT_REPAIRABLE issue (RETIRE or REPLACE).
     */
    MaintenanceIssueResponse applyManagerDecision(Long issueId, ManagerDecisionRequest request);

    /** Get all work orders for a specific issue. */
    List<MaintenanceWorkOrderResponse> getWorkOrdersByIssue(Long issueId);

    /** Get all work orders assigned to a specific technician. */
    List<MaintenanceWorkOrderResponse> getWorkOrdersByTechnician(Long technicianId);

    /** Get complete maintenance history for an asset. */
    MaintenanceHistoryResponse getAssetMaintenanceHistory(Long assetId);
}

File: frontend/src/api/maintenanceService.ts
import api from "./axios";
import type {
  MaintenanceIssueResponse,
  MaintenanceWorkOrderResponse,
  MaintenanceHistoryResponse,
  ReportIssueRequest,
  AssignTechnicianRequest,
  RespondWorkOrderRequest,
  RejectWorkOrderRequest,
  CompleteRepairRequest,
  NotRepairableRequest,
  ManagerDecisionRequest
} from "../types/maintenance";

export const maintenanceService = {
  // Issues
  reportIssue: async (data: ReportIssueRequest): Promise<MaintenanceIssueResponse> => {
    const res = await api.post("/maintenance/issues", data);
    return res.data;
  },
  
  getIssues: async (): Promise<MaintenanceIssueResponse[]> => {
    const res = await api.get("/maintenance/issues");
    return res.data;
  },
  getIssuesByReporter: async (employeeId: number): Promise<MaintenanceIssueResponse[]> => {
    const res = await api.get(`/maintenance/issues/reported-by/${employeeId}`);
    return res.data;
  },
  
  getIssueById: async (id: number): Promise<MaintenanceIssueResponse> => {
    const res = await api.get(`/maintenance/issues/${id}`);
    return res.data;
  },

  assignTechnician: async (issueId: number, data: AssignTechnicianRequest): Promise<MaintenanceWorkOrderResponse> => {
    const res = await api.post(`/maintenance/issues/${issueId}/work-orders`, data);
    return res.data;
  },

  getWorkOrdersByIssue: async (issueId: number): Promise<MaintenanceWorkOrderResponse[]> => {
    const res = await api.get(`/maintenance/issues/${issueId}/work-orders`);
    return res.data;
  },

  applyIssueDecision: async (issueId: number, data: ManagerDecisionRequest): Promise<MaintenanceIssueResponse> => {
    const res = await api.put(`/maintenance/issues/${issueId}/decision`, data);
    return res.data;
  },

  // Work Orders
  startWorkOrder: async (workOrderId: number): Promise<MaintenanceWorkOrderResponse> => {
    const res = await api.put(`/maintenance/work-orders/${workOrderId}/start`);
    return res.data;
  },

  respondToWorkOrder: async (workOrderId: number, data: RespondWorkOrderRequest): Promise<MaintenanceWorkOrderResponse> => {
    const res = await api.put(`/maintenance/work-orders/${workOrderId}/respond`, data);
    return res.data;
  },

  rejectWorkOrder: async (workOrderId: number, data: RejectWorkOrderRequest): Promise<MaintenanceWorkOrderResponse> => {
    const res = await api.put(`/maintenance/work-orders/${workOrderId}/reject`, data);
    return res.data;
  },

  completeRepair: async (workOrderId: number, data: CompleteRepairRequest): Promise<MaintenanceWorkOrderResponse> => {
    const res = await api.put(`/maintenance/work-orders/${workOrderId}/complete`, data);
    return res.data;
  },

  markNotRepairable: async (workOrderId: number, data: NotRepairableRequest): Promise<MaintenanceWorkOrderResponse> => {
    const res = await api.put(`/maintenance/work-orders/${workOrderId}/not-repairable`, data);
    return res.data;
  },

  applyWorkOrderDecision: async (workOrderId: number, data: ManagerDecisionRequest): Promise<MaintenanceWorkOrderResponse> => {
    const res = await api.put(`/maintenance/work-orders/${workOrderId}/decision`, data);
    return res.data;
  },

  getWorkOrdersByTechnician: async (technicianId: number): Promise<MaintenanceWorkOrderResponse[]> => {
    const res = await api.get(`/maintenance/work-orders/technician/${technicianId}`);
    return res.data;
  },

  // History
  getAssetHistory: async (assetId: number): Promise<MaintenanceHistoryResponse> => {
    const res = await api.get(`/maintenance/assets/${assetId}/history`);
    return res.data;
  }
};

